sklearn/manifold/_isomap.py

```diff
@@ -186,9 +192,54 @@ def _fit_transform(self, X):
             n_jobs=self.n_jobs,
         )
 
-        self.dist_matrix_ = graph_shortest_path(
-            kng, method=self.path_method, directed=False
-        )
+        # Compute the number of connected components, and connect the different
+        # components to be able to compute a shortest path between all pairs
+        # of samples in the graph.
+        # Similar fix to cluster._agglomerative._fix_connectivity.
+        n_connected_components, labels = connected_components(kng)
+        if n_connected_components > 1:
+            if self.metric == "precomputed":
+                raise RuntimeError(
+                    "The number of connected components of the neighbors graph"
+                    f" is {n_connected_components} > 1. The graph cannot be "
+                    "completed with metric='precomputed', and Isomap cannot be"
+                    "fitted. Increase the number of neighbors to avoid this "
+                    "issue."
+                )
+            warnings.warn(
+                "The number of connected components of the neighbors graph "
+                f"is {n_connected_components} > 1. Completing the graph to fit"
+                " Isomap might be slow. Increase the number of neighbors to "
+                "avoid this issue.",
+                stacklevel=2,
+            )
+            # use array validated by NearestNeighbors
+            X = self.nbrs_._fit_X
+
+            for i in range(n_connected_components):
+                idx_i = np.where(labels == i)[0]
+                Xi = X[idx_i]
+                for j in range(i):
+                    idx_j = np.where(labels == j)[0]
+                    Xj = X[idx_j]
+                    D = pairwise_distances(
+                        Xi,
+                        Xj,
+                        metric=self.nbrs_.effective_metric_,
+                        **self.nbrs_.effective_metric_params_,
+                    )
+                    ii, jj = np.where(D == np.min(D))
+                    ii, jj = ii[0], jj[0]
```

Comment:
That would not be equivalent to 

```suggestion
                    ii, jj = np.unravel_index(
                        D.argmin(axis=None), D.shape
                    )
```
I think?
Since argmin takes the first occurrence. Should also be faster.